articles/quickstart/native/ionic4/01-login.md

```diff
@@ -0,0 +1,443 @@
+};
+```
+
+### Create an Authentication Service
+
+To coordinate authentication tasks, it's best to set up an injectable service that can be reused across the application. This service needs methods for logging users in and out, as well as checking their authentication state.
+
+```js
+// src/app/services/auth.service.ts
+import { Injectable, NgZone } from '@angular/core';
+import { Storage } from '@ionic/storage';
+import { SafariViewController } from '@ionic-native/safari-view-controller/ngx';
+
+// Import AUTH_CONFIG, Auth0Cordova, and auth0.js
+import { AUTH_CONFIG } from './auth.config';
+import Auth0Cordova from '@auth0/cordova';
+import * as auth0 from 'auth0-js';
+
+declare let cordova: any;
+
+@Injectable()
+export class AuthService {
+  Auth0 = new auth0.WebAuth(AUTH_CONFIG);
+  Client = new Auth0Cordova(AUTH_CONFIG);
+  accessToken: string;
+  user: any;
+  loggedIn: boolean;
+  loading = true;
+
+  constructor(
+    public zone: NgZone,
+    private storage: Storage,
+    private safariViewController: SafariViewController
+  ) {
+    this.storage.get('profile').then(user => this.user = user);
+    this.storage.get('access_token').then(token => this.accessToken = token);
+    this.storage.get('expires_at').then(exp => {
+      this.loggedIn = Date.now() < JSON.parse(exp);
+      this.loading = false;
+    });
+  }
+
+  login() {
+    this.loading = true;
+    const options = {
+      scope: 'openid profile offline_access'
+    };
+    // Authorize login request with Auth0: open login page and get auth results
+    this.Client.authorize(options, (err, authResult) => {
+      if (err) {
+        throw err;
+      }
+      // Set access token
+      this.storage.set('access_token', authResult.accessToken);
+      this.accessToken = authResult.accessToken;
+      // Set access token expiration
+      const expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
+      this.storage.set('expires_at', expiresAt);
+      // Set logged in
+      this.loading = false;
+      this.loggedIn = true;
+      // Fetch user's profile info
+      this.Auth0.client.userInfo(this.accessToken, (err, profile) => {
+        if (err) {
+          throw err;
+        }
+        this.storage.set('profile', profile).then(val =>
+          this.zone.run(() => this.user = profile)
+        );
+      });
+    });
+  }
+
+  logout() {
+    this.storage.remove('profile');
+    this.storage.remove('access_token');
+    this.storage.remove('expires_at');
+    this.accessToken = null;
+    this.user = null;
+    this.loggedIn = false;
+
+    this.safariViewController.isAvailable()
+      .then((available: boolean) => {
+        const domain = AUTH_CONFIG.domain;
+        const clientId = AUTH_CONFIG.clientId;
+        const pkgId = AUTH_CONFIG.packageIdentifier;
+        let url = `https://${domain}/v2/logout?client_id=${clientId}&returnTo=${pkgId}://${domain}/cordova/${pkgId}/callback`;
```

Comment:
`domain` needs to be escaped - https://github.com/auth0/docs/blob/master/CONTRIBUTING.md#escaping-strings